### ISSUE TITLE
General Cleanup & Improvements - DevOps

### ISSUE THREAD

**Issue by george43g** [blockframes/blockframes#3076]
After conducting a number of interviews, I wanted to note down some of the issues that have been discussed, and start formulating proposals to fix them, as well as discussing priority of what should be done first.

This is not an exhaustive list of tasks and is probably going to be a "precursor" to other github issues as this is a broad list.

## Delete/Move folders/files
There are some files that are garbage/unused and should be deleted, and there are some folders which need to be moved.

- `cypress/*` This is a cypress folder that sits in the root of the repo. It has no tests. It should be deleted.
- `docs/` This is an incomplete attempt at documentation. This folder should either be removed, or updated to match documentation on Notion / generated documentation.
- `libs/*` should have a nested structure for clarity. It is possible to use a move schematic to create directories such as `shared`, `testing`, `project-specific` to better organise libraries. For example, `e2e` should be in `libs/testing/cypress/`
- `env/` is suggested removed as part of a separate issue
- `ops/` this folder doubles up on the `tools` folder which is provided for this purpose. Consolidate scripts in `tools/`.
- `styles/` should be a lib
- `tools/credentials/*.json` this directory should store user's service account
- update .gitignore
- `rules/` folder to hold firebase security rules files and indexes (there are migration scripts which depend on the location of these files using hardcoded string literals)
- `deploy-prod.sh` - investigate and move to appropriate dir
- update readme.md with basic quick start (if not generating more docs)

## Other Proposed Tasks

- investigate all *.sh files. By removing the need for shell scripts, we can allow windows devs to not use WSL, which would enable compatibility with Cypress.
- instead of rewriting scripts in js/ts, develop a backend CLI tool used by devs and devops scripts.
- review the strategy used to manage `functions.config()` variables. A script should regularly export these to the root as `.runtimeConfig.json` (check this name)
- review NX tags and lint rules
- review backend-functions structure - utilize method found in `better-firebase-functions`
- review config.yml for CI

## More

- Using Zapier to trigger firestore backups. Backups from prod are stored in a storage bucket, later used for staging data after being anonymised.
- `.firebaserc` related problems - resources are not automatically provisioned, so each dev must make their own addition to this file in order to deploy various resources to their env. Consider using shared development environments and local emulators (which still need isolated environments). Consider adding this file to gitignore, or using firebase-tools programmatically. Investigate multi-tenancy to see if projects can be automatically generated.
- automate generation of test data / golden data (should be faked/generated reliably)
- improve migration workflows (backup scripts, automate)
- improve local deployment and deployment scripts in CI (CLI would help)
- remove secret page in web app that performs admin functions. either use admin app or CLI
- use local scripts where possible rather than "admin" cloud functions. For production backup, use newer google-provided firestore export tools.
- consider more robust solution for storing backup data - not a firebase project.
- firebase functions are disabled via a flag in firestore - improve on this (use config var?)
- firestore is locked via deploying a new ruleset via CLI - investigate if possible to do this faster/programmatically or using a rule - ie - if a document exists, deny ALL writes/reads.
- improve security rules (general improvements from experience in writing these)
- use npm scripts and angular.json as a source of truth for scripts. As much as possible should be contained in the angular.json configuration first - this includes build/test/deploy scripts with various configurations. These can be called from scripts in `npm run *`. This will allow us to significantly simplify the `config.yml` CircleCI file.

### e2e Tests

There are several improvements we can make:

- plugins can be written in TS, utilise plugins to perform seeding / login tasks automatically (significantly reduce e2e time)
- investigate dynamically generating seed data 
- allow automatic use and configuration of emulators, to enable devs to easily perform Test-Driven-Development with cypress running on any platform

## Git / Changelog

- enforce `conventional-changelog`, commitizen, `semantic-release` for automatic changelog generation and semantic versioning based on commit history in CI
- use `husky` to enforce git hooks.

This is not a complete list as of 15/6

**Comment by GrandSchtroumpf** [comment]
Some quick feedback : 
- "`libs/*` should have a nested structure for clarity": We started with that and it makes things super difficult to maintain. Libraries like `shared` become garbage libraries even with a lot of effort to maintain them clean (take a look at `ui` & `utils`). Also it makes thinks harder to find out. We are moving toward an opposite direction with a flatter architecture.
- "`styles/` should be a lib": do you have any strong argument for that (other than reducing the amount of root folders) ?
- "`rules/` folder to hold firebase security rules files and indexes": If the folder holds indexes & rules it should be be called `firestore`. Why this one should be at the root while other root folders should be in libs ?
- "develop a backend CLI tool": I thought you wanted to use npm scripts to have a single source of actions.
- "consider more robust solution for storing backup data - not a firebase project.": If firebase project is not rebust enough, then we can stop working right now ;)
- "allow automatic use and configuration of emulators": I think this is the most important job right now.

**Comment by phpgeekfr** [comment]
@george43g can we close or update this issue ?

**Comment by george43g** [comment]
@phpgeekfr the intention here was to scan through and find anything leftover or important that we need and distil it into issues in the main devops issue summary, and then close this. I haven't had a chance to do this yet.
